jetty-server/src/main/java/org/eclipse/jetty/server/DetectorConnectionFactory.java

```diff
@@ -0,0 +1,293 @@
+{
+    private static final Logger LOG = Log.getLogger(DetectorConnectionFactory.class);
+
+    protected final List<Detecting> _detectingConnectionFactories;
+
+    /**
+     * <p>When the first bytes are not recognized by the {@code detectingConnectionFactories}, the default behavior is to
+     * upgrade to the protocol returned by {@link #findNextProtocol(Connector)}.</p>
+     * @param detectingConnectionFactories the {@link Detecting} instances.
+     */
+    public DetectorConnectionFactory(Detecting... detectingConnectionFactories)
+    {
+        super(toProtocolString(detectingConnectionFactories));
+        _detectingConnectionFactories = Arrays.asList(detectingConnectionFactories);
+        for (Detecting detectingConnectionFactory : detectingConnectionFactories)
+        {
+            addBean(detectingConnectionFactory);
+        }
+    }
+
+    static String toProtocolString(Detecting... detectingConnectionFactories)
+    {
+        if (detectingConnectionFactories.length == 0)
+            throw new IllegalStateException("At least one detecting instance is required");
+
+        // remove protocol duplicates while keeping their ordering -> use LinkedHashSet
+        LinkedHashSet<String> protocols = new LinkedHashSet<>();
+        for (Detecting dcf : detectingConnectionFactories)
+        {
+            protocols.add(dcf.getProtocol());
+        }
+
+        StringBuilder sb = new StringBuilder();
+        for (String protocol : protocols)
+        {
+            sb.append(protocol).append('|');
+        }
+        sb.deleteCharAt(sb.length() - 1);
+        String protocol = sb.toString();
+        LOG.debug("Detector generated protocol name : {}", protocol);
+        return protocol;
+    }
+
+    /**
+     * Perform a detection using multiple {@code detectingConnectionFactories} and return the aggregated outcome.
+     * @param buffer the buffer to perform a detection against.
+     * @param detectingConnectionFactories the {@link Detecting} instances.
+     * @return A {@link Detecting.Detection} value with the detection outcome of the {@code detectingConnectionFactories}.
+     */
+    protected static Detecting.Detection detectUsing(ByteBuffer buffer, Iterable<Detecting> detectingConnectionFactories)
+    {
+        LOG.debug("Attempting detection from buffer {} using {}", buffer, detectingConnectionFactories);
+        boolean allDetectionNotRecognized = true;
+        for (Detecting detectingConnectionFactory : detectingConnectionFactories)
+        {
+            Detecting.Detection detection = detectingConnectionFactory.detect(buffer);
+            if (detection == Detecting.Detection.RECOGNIZED)
+            {
+                LOG.debug("Detection recognized bytes from buffer {} using {}", buffer, detectingConnectionFactories);
+                return Detecting.Detection.RECOGNIZED;
+            }
+            allDetectionNotRecognized &= detection == Detecting.Detection.NOT_RECOGNIZED;
```

Comment:
Once a factory says not recognized, should it be removed from the list for further iterations?

Reply:
It's an optimization we could add, again at the cost of a slightly more complex implementation.

In practice, as long as the `DetectorConnectionFactory` isn't initialized with dozens of `Detecting` instances and only a small number of bytes are required to perform a detection this shouldn't make a difference IMHO.

Reply:
I'm fine with not doing this optimisation now, but are the method signatures structured in such a way it can be done later without needing to break API?

Reply:
I'm positive the API wouldn't have to change to implement that optimization.